OpenTissue/gpu/image/io/image_read.h

```diff
@@ -1,91 +1,180 @@
-      
-      OpenTissue::image::detail::ilImage devil;
-
-      ILboolean result = devil.Load(const_cast< char * >(filename.c_str()));
-      if (!result)
-      {
-        return false;
-      }
-
-      if (show_statistics)
-      {
-        std::cout << "--- file : " << filename << "---------------------------" << std::endl;
-        std::cout << "\twidth         = " << (int)devil.Width() << std::endl;
-        std::cout << "\theight        = " << (int)devil.Height() << std::endl;
-        std::cout << "\tdepth         = " << (int)devil.Depth() << std::endl;
-        std::cout << "\tBpp           = " << (int)devil.Bpp() << std::endl;
-        std::cout << "\tBitpp         = " << (int)devil.Bitpp() << std::endl;
-        std::cout << "\tPallette Type = " << OpenTissue::image::detail::get_IL_string(devil.PaletteType()) << std::endl;
-        std::cout << "\tFormat        = " << OpenTissue::image::detail::get_IL_string(devil.Format()) << std::endl;
-        std::cout << "\tType          = " << OpenTissue::image::detail::get_IL_string(devil.Type()) << std::endl;
-        std::cout << "\tNumImages     = " << (int)devil.NumImages() << std::endl;
-        std::cout << "\tNumMipmaps    = " << (int)devil.NumMipmaps() << std::endl;
-        std::cout << "\tOrigin        = " << OpenTissue::image::detail::get_IL_string(devil.GetOrigin()) << std::endl;
-      }
-
-      if (devil.GetOrigin() != IL_ORIGIN_LOWER_LEFT)
-        devil.Flip();
-
-      size_t width  = devil.Width();
-      size_t height = devil.Height();
-      image.create(width, height, 4);
-
-      devil.Bind();
-      ilCopyPixels(0, 0, 0, width, height, 1, IL_RGBA, IL_UNSIGNED_BYTE, image.get_data() );
-
-      ilShutDown();
-
-      return true;
-    }
-
-  } // namespace image
-} // namespace OpenTissue
+      std::cerr << "error: failed to open png file " << filename << "." << std::endl;
+      return false;
+  }
+
+  // Read the header
+  png_byte header[8];
+  fread(header, 1, 8, fp);
+
+  if (png_sig_cmp(header, 0, 8))
+  {
+      std::cerr << "error: " << filename << " is not a PNG." << std::endl;
+      fclose(fp);
+      return false;
+  }
+
+  png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+  if(!png)
+  {
+    std::cerr << "error: png_create_read_struct returned 0." << std::endl;
+    fclose(fp);
+    return false;
+  }
+
+  png_infop info = png_create_info_struct(png);
+  if(!info)
+  {
+    std::cerr << "error: png_create_info_struct returned 0." << std::endl;
+    png_destroy_read_struct(&png, nullptr, nullptr);
+    fclose(fp);
+    return false;
+  }
+
+  // Create png end info struct
+  png_infop end_info = png_create_info_struct(png);
+  if (!end_info)
+  {
+      std::cerr << "error: png_create_info_struct returned 0" << std::endl;
+      png_destroy_read_struct(&png, &info, nullptr);
+      fclose(fp);
+      return false;
+  }
+
+
+  if (setjmp(png_jmpbuf(png))) {
```

Comment:
Inconsistent formatting of curly braces

Reply:
hmm, are you reviewing my the first commit on this PR? this is fixed on the second commit.